# This patch file was generated by NetBeans IDE
# Following Index: paths are relative to: D:\Code\SVN\trunk\China_address_field
# This patch can be applied using context Tools: Patch action on respective folder.
# It uses platform neutral UTF-8 encoding and \n newlines.
# Above lines and this line are ignored by the patching process.
Index: china_address.module
--- china_address.module 基本 (BASE)
+++ china_address.module 本地修改 (基于LOCAL)
@@ -139,53 +139,41 @@
   $lang = 'und';
   switch ($instance['widget']['type']) {
     case 'china_address_default':
+      //需要将字段信息传递到表单其它地方
+      $form_state['china_address'] = array('field' => $field);
       // Set default value.
       $province = 2;
       $city = 52;
       $county = 500;
-
-      // Province.
-      if ($field['settings']['china_address_level'] == CHINA_ADDRESS_CITY_COUNTY && $field['settings']['china_address_province_default']) {
-        $province = $field['settings']['china_address_province_default'];
+      $detail = '';
+      if(!empty($form_state['triggering_element']['#china_address']) && !empty($form_state['values'])){
+          $addr_values = $form_state['values'];
+          foreach($form['#parents'] as $child){
+              if(is_array($addr_values) && array_key_exists($child, $addr_values)){
+                  $addr_values = $addr_values[$child];
       }
-      elseif (isset($form_state['values'][$field['field_name']][$lang][0]['province'])) {
-        // Ajax callback.
-        $province = $form_state['values'][$field['field_name']][$lang][0]['province'];
       }
-      elseif (isset($items[$delta]['province'])) {
-        $province = $items[$delta]['province'];
+          $addr_values = $addr_values[$field['field_name']][$langcode][0];
+      }else{
+          $addr_values = !empty($items)? $items[$delta]: array();
       }
+      $province = empty($addr_values['province'])? $province: $addr_values['province'];
+      $city = empty($addr_values['city'])? $province: $addr_values['city'];
+      $county = empty($addr_values['county'])? $province: $addr_values['county'];
+      $detail = empty($addr_values['detail'])? $detail: $addr_values['detail'];
 
-      // City.
-      if (isset($form_state['values'][$field['field_name']][$lang][0]['city'])) {
-        // Ajax callback.
-        $city = $form_state['values'][$field['field_name']][$lang][0]['city'];
-      }
-      elseif (isset($items[$delta]['city'])) {
-        $city = $items[$delta]['city'];
-      }
       $cities = china_address_cities($province);
       $city_keys = array_keys($cities);
       if (!in_array($city, $city_keys)) {
         $city = $city_keys[0];
       }
-
-      // County.
-      if (isset($form_state['values'][$field['field_name']][$lang][0]['county'])) {
-        // Ajax callback.
-        $county = $form_state['values'][$field['field_name']][$lang][0]['county'];
-      }
-      elseif (isset($items[$delta]['county'])) {
-        $county = $items[$delta]['county'];
-      }
       $counties = china_address_counties($city);
       $county_keys = array_keys($counties);
       if (!in_array($county, $county_keys)) {
         $county = $county_keys[0];
       }
+      setcookie('address', $province.'-'.$city.'-'.$county,time() + 3600 ,'/');
 
-      // Detail.
-      $detail = isset($items[$delta]['detail']) ? $items[$delta]['detail'] : '';
 
       $element['#prefix'] = '<div id="china-address-dropdown-replace-' . str_replace('_', '-', $field['field_name']) . '" class="china-address-dropdown-replace" >';
       $element['#suffix'] = '</div>';
@@ -203,6 +191,7 @@
           'callback' => 'china_address_dropdown_callback',
           'wrapper' => 'china-address-dropdown-replace-' . str_replace('_', '-', $field['field_name']),
         ),
+        '#china_address' => TRUE //唯一识别
       );
       $element['province'] = array(
         '#type' => 'select',
@@ -256,7 +245,7 @@
 function china_address_provinces($province = NULL) {
   $provinces = china_address_fetch_region(1, 1);
   if ($province) {
-    return isset($provinces[$province]) ? $provinces[$province] : '';
+    return isset($provinces[$province]) ? $provinces[$province] : array();
   }
   return $provinces;
 }
@@ -267,7 +256,7 @@
 function china_address_cities($parent_id, $city = NULL) {
   $cities = china_address_fetch_region($parent_id, 2);
   if ($city) {
-    return isset($cities[$city]) ? $cities[$city] : '';
+    return isset($cities[$city]) ? $cities[$city] : array();
   }
   return $cities;
 }
@@ -278,7 +267,7 @@
 function china_address_counties($parent_id, $county = NULL) {
   $counties = china_address_fetch_region($parent_id, 3);
   if ($county) {
-    return isset($counties[$county]) ? $counties[$county] : '';
+    return isset($counties[$county]) ? $counties[$county] : array();
   }
   return $counties;
 }
@@ -310,17 +299,21 @@
 /**
  * Selects dropdown to be returned for re-rendering.
  */
-function china_address_dropdown_callback($form, $form_state) {
-  // $lang = isset($form['language']['#value']) ?
-  // $form['language']['#value'] : 'und';
-  $lang = 'und';
-  if ($form['#form_id'] == 'field_ui_field_edit_form') {
-    return $form['instance']['default_value_widget'][$form['#field']['field_name']][$lang][0];
+function china_address_dropdown_callback(&$form, &$form_state) {
+    $field_name = $form_state['china_address']['field']['field_name'];
+    $field = $form;
+    if(in_array($field_name, $form_state['triggering_element']['#parents'])){
+        foreach($form_state['triggering_element']['#parents'] as $child){
+            if($child != $field_name){
+                $field = $field[$child];
+            }else{
+                break;
   }
-  else {
-    return $form[$form_state['triggering_element']['#parents'][0]][$lang][0];
   }
 }
+    $field = $field[$field_name];
+    return $field;
+}
 
 /**
  * Implements hook_field_formatter_info().
